FlintCore/Utils/CompletionRequirement.swift

```diff
@@ -0,0 +1,210 @@
+/// if the completion block has been called asynchronously.
+///
+/// // Call the function that requires completion
+/// let status = coordinator.doSomething(input: x, completionRequirement: completion)
+///
+/// // Make sure one of the valid statuses was returned.
+/// // This safety test ensures that the completion from the correct completion requirement instance was returned.
+/// precondition(completion.verify(status))
+///
+/// // If the result does not return true for `isCompletingAsync`, the completion callback will have already been called by now.
+/// if !status.isCompletingAsync {
+///     print("Completed synchronously: \(status.value)")
+/// } else {
+///     print("Completing asynchronously... see you later")
+/// }
+/// ```
+///
+/// When implementing such a function requiring a completion handler, you return one of two statuses returned by either
+/// the `CompletionRequirement.completed(_ arg: T)` or `CompletionRequirement.willCompleteAsync()`.
+/// The `CompletionRequirement` will take care of calling the completion block as appropriate.
+///
+/// ```
+/// func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status {
+///     return completionRequirement.completedSync(false)
+/// }
+///
+/// // or for async completion
+///
+/// func doSomething(input: Any, completionRequirement: DoSomethingCompletion) -> DoSomethingCompletion.Status {
+///     let result = completionRequirement.willCompleteAsync()
+///     DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
+///         result.completed(false)
+///     }
+///     return result
+/// }
+/// ```
+public class CompletionRequirement<T> {
+    public class Status {
+        fileprivate let _value: T?
+        fileprivate var value: T {
+            get {
+                guard let result = _value else {
+                    flintBug("CompletionRequirement status value is nil")
+                }
+                return result
+            }
+        }
+        
+        init() {
+            _value = nil
+        }
+
+        init(result: T) {
+            _value = result
+        }
+        
+        public var isCompletingAsync: Bool { return false }
+    }
+
+    // The type for a status indicating completion will occur later
+    public class DeferredStatus: Status {
+        var owner: CompletionRequirement<T>?
+        
+        init(owner: CompletionRequirement<T>) {
+            self.owner = owner
+            super.init()
+        }
+
+        override public var isCompletingAsync: Bool {
+            return true
+        }
+        
+        public func completed(_ result: T) {
+            guard let owner = owner else {
+                return
+            }
+            owner.completionHandler(result, true)
+        }
+    }
+
+    fileprivate var completionStatus: Status?
+    var completionHandler: ((T, _ completedAsync: Bool) -> Void)!
```

Comment:
Instead of an IUO (implicitly unwrapped optional), can you make it optional and do a fatalError so at least you have info as to what the issue was? If this is ever unwrapped an optional it will only say "unexpectedly found nil" and it will be tough to track down.

Reply:
I'm happy to do this, I am reworking this slightly anyway to support specifying the queue on which to call it - but a stacktrace would always show where it occurred so right now I don't totally buy into the IUO hate for values that *must* always initialised but cannot always be set in the initialiser — note the `init()` that leaves `completionHandler` unset is `fileprivate` and only used by the proxy completion (ugh that was some ugly stuff to get working!)